### DISCUSSION TITLE
Help with Programming an SPI Flash via the Bootloader on Arty A7-100 Rev.E (spansion)?

### DISCUSSION THREAD

**Discussion by jmstein7** [stnolting/neorv32#211] [Q&A]
Stephan,

First off, I _love_ your project! I'm fairly new to RISC V and FPGAs, but you've made this _so_ easy to understand, as well as providing good performance. But, I digress...

I've been experimenting with various neorv32 SoCs on my Arty A7-100 Rev.E, but I just cannot get "store to flash" or "load from flash," _vis a vis_ the bootloader, to work with the Spansion  S25FL128SAGMF100 (16 MB SPI flash). I tried updating the commands to the 4 bit commands in the datasheet, but, nothing. It just hangs. Everything else works - I can "upload" and "execute" my programs with ease. But, no matter how I try, I cannot program the External SPI Flash via the Bootloader.

One thing I noticed is that the bootloader looks to 0x0800_0000, though SPI seems to start in the peripheral space between 0xFFFFE00 and 0xFFFFFFF.

So, color me confused! Any help you could offer would be greatly appreciated!

Jon

**Comment by stnolting** [comment]
Hey there!

Thank you very much for the nice words :)

I think the problem is just an address mapping issue.
The bootloader expects the executable in the flash at the address defined by `SPI_BOOT_BASE_ADDR`. By default this is `0x08000000`, so the executable is located at the flash's 128MB boundary. Obviously, this is out of range for your setup. You could try setting the address to 8MB (0x00800000). This requires re-compiling the bootloader:

`sw/bootloader$ make clean_all USER_FLAGS+=-DSPI_BOOT_BASE_ADDR=0x00800000 bootloader`

**Reply by jmstein7** [reply]
Stephan,

Thanks for the reply. I tried changing the address around, but no luck. It's 16mb, though, not 128.

It says: 

Write 0x00003fec bytes to SPI flash @0x0x00600000? (y/n) y
Flashing...

Then, it hangs.

Anything else you could suggest?

I put my SoC implementation here:  https://github.com/jmstein7/neorv32_arty_A7

**Reply by stnolting** [reply]
Then there is a problem in the SPI access functions. 🤔 Unfortunately, the bootloader does not provide a lot of error messages because it is optimized for size. Did you try running the original bootloader code with the modified base address and without any code changes?

I just had a quick look at the data sheet of your SPI flash. Seems like all base commands are supported, except for `SPI_FLASH_CMD_READ_ID` which is `0x9E` in the default bootloader implementation but the flash requires `0x9f`. You could try changing that, recompile the bootloader and re-synthesized the design.

**Reply by jmstein7** [reply]
Yes, I tried running the original code with the new base address but without any code changes - and then again with 0x9F for read id. It didn't work. I would have left it alone and just given up, but for the fact that the Arty A7 is a popular platform for this precise kind of experimentation (with RISC-V in general and with your neorv32 core in particular). So, there is value in figuring this out beyond just my little SoC. I think many others will try, fail, and conclude that either 1) it's broken, 2) they broke something, or 3) there is some problem with the board.

**Reply by stnolting** [reply]
I absolutely agree 😉 
What happens if you type `l` (load rom flash) in the bootloader console? What error message does it output?

**Reply by jmstein7** [reply]
It returns: 
ERROR_0

**Reply by stnolting** [reply]
That's not bad! At least the SPI read functions seem to work. ERROR_0 means that the data at the specified base address does not contain a valid executable. If there was a problem accessing the flash the bootloader would have printed ERROR_3.

You could try setting the base address `SPI_BOOT_BASE_ADDR` to all zero `0x00000000` - maybe there is a problem is byte/word/page addressing. This will override any FPGA bitstream, but that should not be a problem.

**Reply by jmstein7** [reply]
So, I got the “good” error. 😁

I’ll try jiggering the addresses around again, but I did try the suggested offset of 0x0060_0000, which is the end of any bitstream stored in the flash. 

Between the Flash and the DDR3L, Xilinx sure does make it hard to make full use of the A7’s capabilities and features.

**Reply by stnolting** [reply]
I have added a new example program [`sw/example/demo_spi`](https://github.com/stnolting/neorv32/tree/master/sw/example/demo_spi) that can be used to create SPI transactions "by hand" via a simple UART console. You could try using that to check if you can access your flash memory at all.

For example:
* set wrtie-enable latch (command `0x06`)
* check if WE flag is set in status register (command `0x05`)
* check the address size (should be 24 bits for the bootloader)
* ...

**Reply by jmstein7** [reply]
I shall! What a fantastic debugging tool! 

I'm going to try it with some microchip 23LC512s too (I designed some simple, custom PMODs for those). 

Jon

**Comment by stnolting** [comment]
Maybe this is caused by a signal mapping problem. The default clock flash clock signal cannot be used as general purpose IO, but there is an additional pin connected to the flash that has to be used (L16). This is what my pin mapping for the Arty A-35 looks like:
```
set_property -dict { PACKAGE_PIN L13   IOSTANDARD LVCMOS33 } [get_ports { spi_csn_o }];
set_property -dict { PACKAGE_PIN K17   IOSTANDARD LVCMOS33 } [get_ports { spi_sdo_o }];
set_property -dict { PACKAGE_PIN K18   IOSTANDARD LVCMOS33 } [get_ports { spi_sdi_i }];
set_property -dict { PACKAGE_PIN L16   IOSTANDARD LVCMOS33 } [get_ports { spi_sck_o }];
```

I think the pin mapping is the same for the A-100.

**Reply by jmstein7** [reply]
I see we're on the "same page" here, lol! This is what I have:

set_property PACKAGE_PIN L16 [get_ports spi_sck_o]
set_property PACKAGE_PIN K17 [get_ports spi_sdi_i]
set_property PACKAGE_PIN K18 [get_ports spi_sdo_o]
set_property PACKAGE_PIN L13 [get_ports {spi_csno[0]}]

set_property IOSTANDARD LVCMOS33 [get_ports {spi_csno[0]}]
set_property IOSTANDARD LVCMOS33 [get_ports spi_sck_o]
set_property IOSTANDARD LVCMOS33 [get_ports spi_sdi_i]
set_property IOSTANDARD LVCMOS33 [get_ports spi_sdo_o]

**Comment by Unike267** [comment]
Hi there!

> I have added a new example program [sw/example/demo_spi](https://github.com/stnolting/neorv32/tree/master/sw/example/demo_spi)

I'm using a modification of `demo_spi` program because in CuteCom terminal the `neorv32_uart0_scan` doesn't work fine and when request a value the number stays freeze in the terminal and doesn't enter to the software.  Therefore, I'm using the bootloader functions `neorv32_uart0_getc()` `neorv32_uart0_putc(a)` and with these functions (in CuteCom terminal) the program works fine. But that is not the issue.

The issue is that when I make a SPI data transmission I don't receive a response from the flash.

I've set a ILA for make a debug and I've checked that the flash is enable and the data is sent correctly.

As is shown in the following image:

![SPI-ILA](https://github.com/stnolting/neorv32/assets/36209778/1929c79a-ed8d-4170-be2c-1caf653f276e)

According to the [datasheet](https://pdf1.alldatasheet.com/datasheet-pdf/view/353330/NUMONYX/N25Q128A13EF840E.html) when code `0x9F` is sent the SPI Flash should respond with the Manufacturer identification. But there isn't answer.

Moreover, it doesn't respond for any code. Could you give me any hint?

> set_property -dict { PACKAGE_PIN L13   IOSTANDARD LVCMOS33 } [get_ports { spi_csn_o }];
set_property -dict { PACKAGE_PIN K17   IOSTANDARD LVCMOS33 } [get_ports { spi_sdo_o }];
set_property -dict { PACKAGE_PIN K18   IOSTANDARD LVCMOS33 } [get_ports { spi_sdi_i }];
set_property -dict { PACKAGE_PIN L16   IOSTANDARD LVCMOS33 } [get_ports { spi_sck_o }];

I have this constrains in the `.xdc`.

Thanks!

---

/cc @umarcor 

--- 

Edit:

Although this would not have worked, because as you can see below the problem was that I enabled the chip before doing the setup. The ILA image is wrong because I didn't send a 0x00 sequence to send to the flash-spi memory the clock and get a result.

**Reply by stnolting** [reply]
> I'm using a modification of demo_spi program because in CuteCom terminal the neorv32_uart0_scan doesn't work fine and when request a value the number stays freeze in the terminal and doesn't enter to the software. Therefore, I'm using the bootloader functions neorv32_uart0_getc() neorv32_uart0_putc(a) and with these functions (in CuteCom terminal) the program works fine. But that is not the issue.

Can you try another terminal program? Maybe HTerm or TeraTerm?

> Moreover, it doesn't respond for any code. Could you give me any hint?

How did you configure the SPI module? Maybe the clock mode does not match or the SPI frequency is too high. I just tested to access the flash on my DE0-nano board, which looks good:

```
<<< SPI Bus Explorer >>>

This program allows to create SPI transfers by hand.
Type 'help' to see the help menu.

SPI_EXPLORER:> setup
Select SPI clock prescaler (0..7): 7
Enter clock divider (0..15, as one hex char): 0
+ New SPI clock speed = 12207 Hz
Select SPI clock mode (0..3): 0
+ New SPI clock mode = 0

SPI_EXPLORER:> en
Chip-select line to ENABLE (set low) [0..7]: 0
SPI_EXPLORER:> trans
Enter TX data (2 hex chars): 0x9f
TX data: 0x9f
RX data: 0xff
SPI_EXPLORER:> trans
Enter TX data (2 hex chars): 0x00
TX data: 0x00
RX data: 0x20
SPI_EXPLORER:> trans
Enter TX data (2 hex chars): 0x00
TX data: 0x00
RX data: 0x20
SPI_EXPLORER:> trans
Enter TX data (2 hex chars): 0x00
TX data: 0x00
RX data: 0x17
SPI_EXPLORER:> trans
Enter TX data (2 hex chars): 0x00
TX data: 0x00
RX data: 0x10
SPI_EXPLORER:> trans
Enter TX data (2 hex chars): 0x00
TX data: 0x00
RX data: 0x00
SPI_EXPLORER:> dis
Disabling chip select lines.
SPI_EXPLORER:>
```

**Reply by Unike267** [reply]
Hi Stephan!

> Can you try another terminal program? Maybe HTerm or TeraTerm?

I've tried with the **picocom** terminal using the following command `picocom /dev/ttyUSB1 -b 19200 --send-cmd="ascii-xfr -s -n"` (with `ctrl+a+s` to send the program) and the `demo_spi` program works fine. It will be a question of the CuteCom terminal.

> How did you configure the SPI module? Maybe the clock mode does not match or the SPI frequency is too high. I just tested to access the flash on my DE0-nano board, which looks good:

The problem is solved. I've tried with all possible prescaler in `Arty A7` and all work fine.  The following image shows with the prescaler equal to 0 (SPI clock speed = 25MHz):

![spi_flash_trans_ok](https://github.com/stnolting/neorv32/assets/36209778/f5e62740-fbab-4e5c-a64e-875d3c8768fe)

With this configuration (25Mhz) it should work because the [datasheet](https://pdf1.alldatasheet.com/datasheet-pdf/view/353330/NUMONYX/N25Q128A13EF840E.html) indicates that the maximum clock frequency of the flash is 108 MHz.

You won't believe what the problem was, I set the enable before make the setup. And if you don't follow the process in the right order it doesn't work. :sweat_smile: 

As shown in the following image:

![spi_flash_trans_fail](https://github.com/stnolting/neorv32/assets/36209778/a4494e9f-da0a-4e1e-9e81-bc28eb32d97a)

Thank you very much for the quickly feedback! :smile:

**Reply by stnolting** [reply]
Great to see that your setup is working! :)

> You won't believe what the problem was, I set the enable before make the setup. And if you don't follow the process in the right order it doesn't work. 😅

That's a "bug" in the program... 🙈 Thanks for finding that! I've just fixed that: https://github.com/stnolting/neorv32/commit/f4a6365e3827f52dab890f14f422fc54836e1bbb

> Thank you very much for the quickly feedback!

You're welcome.
